index.js

```diff
@@ -1,33 +1,47 @@
-const { extname } = require('path')
 const fs = require('fs').promises
 const { compileTemplate } = require('@vue/compiler-sfc')
 const { optimize: optimizeSvg } = require('svgo')
 
 module.exports = function svgLoader (options = {}) {
   const { svgoConfig, svgo } = options
 
+  const svgRegex = /\.svg(\?(raw|url|component))?$/
   return {
     name: 'svg-loader',
     enforce: 'pre',
-
+    resolveid (id) {
+      return id.match(svgRegex) ? id : null
+    },
     async load (id) {
-      const [path, parameter] = id.split('?')
-
-      if (!extname(path).startsWith('.svg') || parameter === 'url') {
-        return null
+      if (!id.match(svgRegex)) {
+        return
       }
-
-      const svg = await fs.readFile(path, 'utf-8')
-
-      const optimizedSvg = svgo === false ? svg : optimizeSvg(svg, svgoConfig).data
-
-      const { code } = compileTemplate({
-        id: JSON.stringify(id),
-        source: optimizedSvg,
-        transformAssetUrls: false
-      })
-
-      return `${code}\nexport default render`
+      const [path, query] = id.split('?', 2)
+      if (query === 'url') {
+        return path
+      } else if (query === 'raw') {
+        return await fs.readFile(path, 'utf-8')
+      }
+    },
+    async transform (src, id) {
+      if (!id.match(svgRegex)) {
+        return
+      }
+      const [path, query] = id.split('?', 2)
+      if (query === 'component') {
+        const svg = await fs.readFile(path, 'utf-8')
+        const optimizedSvg = svgo === false ? svg : optimizeSvg(svg, svgoConfig).data
+
+        let { code, ast, map } = compileTemplate({
+          id: JSON.stringify(id),
+          source: optimizedSvg,
+          filename: path,
+          transformAssetUrls: false
+        })
+        code = code.replace('export function render', 'function render')
```

Comment:
Why is this line needed?

Reply:
It is not necessary. But I believe less export would be better.

Reply:
Agreed, but maybe then write it like this?
```js
code = code.replace('export function render', 'export default function render')
return { code, ast, map }
```

Reply:
OK. In addition, I found that this line will raise an error during build. Because the type is error. 

`transform` need return `{map?: string | SourceMap, ast? : ESTree.Program}` 
`compileTemplate` provide `{ast?: RootNode, map?: RawSourceMap}`

I don‘t know how to resolve it, so I just remove the `ast` and `map` .